sympy/solvers/simplex.py

```diff
@@ -0,0 +1,821 @@
+
+    >>> lp(min, x, [y + x >= -3])  # same as lp(min, x, [x >= -3])
+    (-3, {x: -3})
+    """
+    A, B, C, D, r, xx, aux = _lp_matrices(f, constr)
+
+    how = str(min_max).lower()
+    if 'max' in how:
+        # _simplex minimizes for Ax <= B so we
+        # have to change the sign of the function
+        # and negate the optimal value returned
+        _o, p, d = _simplex(A, B, -C, -D)
+        o = -_o
+    elif 'min' in how:
+        o, p, d = _simplex(A, B, C, D)
+    else:
+        raise ValueError('expecting min or max')
+
+    # restore original variables and remove aux from p
+    p = dict(zip(xx, p))
+    if r:  # p has original symbols and auxilliary symbols
+        # if r has x: x - z1 use values from p to update
+        r = {k: v.xreplace(p) for k, v in r.items()}
+        # then use the actual value of x (= x - z1) in p
+        p.update(r)
+        # and keep only non-aux variables
+        p = {k: v for k, v in p.items() if k not in aux}
+
+    # not returning dual since there may be extra constraints
+    # when a variable has finite bounds
+    return o, p
+
+
+def _lp_args(min_max, *args):
+    """return standard input for lp from how and args. The
+    method ``how`` is interpreted as being either min or max,
+    and args (up to 4) are interpreted and returned as an
+    object function and list of appropriate inequalities.
+
+    1 arg: Matrix/list
+    2 args: Expr, list of constraints
+    3 args: 3 lists/matrices: A, B, C, D=0
+    4 args: 4 lists/matrices: A, B, C, D
+
+    return: ('max' or 'min', objective, constraints)
+    """
+    how = str(min_max).lower()
+    if 'max' in how:
+        how = 'max'
+    elif 'min' in how:
+        how = 'min'
+    else:
+        raise ValueError('expecting min or max for `how`')
+    if not args or len(args) > 4:
+        raise ValueError('expecting 1 - 4 args')
+    LP = [sympify(i) for i in args]
+    if len(LP) > 2 and all(isinstance(i, (list, MatrixBase)) for i in LP):
+        if len(LP) == 3:
+            LP.append([0])
+        a, b, c, d = [Matrix(i) for i in LP]
+        LP = [Matrix([[a, b], [c, d]])]
+    if len(LP) == 1:
+        m = Matrix(LP[0])
+        if how == 'min':
+            p = primal_dual(m)[0]
+        elif how == 'max':
+            p = primal_dual(m.T)[1]
+        return (how,) + p
+    if len(LP) == 2:
+        LP.append([])
+    f, ineq, unbound = LP
+    if not isinstance(f, Expr):
+        raise ValueError('expecting Expr')
+    if not isinstance(ineq, list):
+        raise ValueError('expecting list of inequalities')
+    if not isinstance(unbound, list):
+        raise ValueError('expecting list of unbound symbols')
+    return how, f, ineq
+
+
+def lpmin(*args):
+    """return minimization of linear equation ``f = c*x - d``
+    under constraints ``A*x >= b``.
+
+    Input can be the ``Matrix([[A, b], [c, d]])``, the individual
+    matrices ``A, b, c, d`` or the function ``f`` and a list of
+    constraints expressed using Ge, Le or Eq.
```

Comment:
I'm not sure if I like the idea that this can take multiple types of input. I think this might make it a little confusing to use. Are there other functions in SymPy that can do something like this?

Reply:
python `range` does this. SymPy often works with `*args` and `args` and handles conversion. @oscarbenjamin point out that "Mathematica's LinearOptimisation function has the signature LinearOptimisation(objective, constraints, vars) but can alternatively accept matrices like LinearOptimisation(f_vec, (c_mat, b_vec))." When the logic is straightforward for dealing with the different inputs, I prefer the single interface. But if sentiment is strongly against it, we could do `lpmin_from_matrix`, `lpmin_from_matrices` in addition to `lpmin` But this will not make the viewed code any more transparent to someone that knows the different outputs: it will be `lpmin(M)` (one arg), `lpmin(f, constr)` (two args) or `lpmin(a, b, c)` or `lpmin(a,p,c,d)`.

Reply:
I think it's reasonable to it this way then. Having multiple functions for the same thing does seem annoying. 

I would maybe mention that the input can be a list constraints expressed using Ge, Le or Eq first, before the matrix stuff. I think most people would prefer not to use the matrix. 

What's the reasoning for accepting either M or A, b, c, d? In what situation would someone prefer to use one or the other. It seems simpler to me to either accept just M or A, b, c, d.

Reply:
If someone is making A,b,c,d from linear_eq_to_matrix then it's not too hard to wrap it up before sending it `lpmin(Matrix([[A,b],[c,d]]))` but I got that wrong enough times that I just didn't want to think about it -- I know I have A,b,c,d right and just want the function to put them in M correctly.